dot/rpc/modules/system_test.go

```diff
@@ -2,3 +2,589 @@
+
+	mockCoreAPIErr := new(mocks.CoreAPI)
+	mockCoreAPIErr.On("GetMetadata", (*common.Hash)(nil)).Return(nil, errors.New("getMetadata error"))
+
+	// Magic number mismatch
+	mockCoreAPIMagicNumMismatch := new(mocks.CoreAPI)
+	mockCoreAPIMagicNumMismatch.On("GetMetadata", (*common.Hash)(nil)).Return(storageKeyHex, nil)
+
+	mockStorageAPI := new(mocks.StorageAPI)
+	mockStorageAPI.On("GetStorage", (*common.Hash)(nil), storageKeyHex).Return(common.MustHexToBytes("0x03000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"), nil)
+
+	mockStorageAPIErr := new(mocks.StorageAPI)
+	mockStorageAPIErr.On("GetStorage", (*common.Hash)(nil), storageKeyHex).Return(nil, errors.New("getStorage error"))
+
+	type args struct {
+		r   *http.Request
+		req *StringRequest
+	}
+	tests := []struct {
+		name      string
+		sysModule *SystemModule
+		args      args
+		expErr    error
+		exp       U64Response
+	}{
+		{
+			name:      "Nil Request",
+			sysModule: NewSystemModule(nil, nil, mockCoreAPI, mockStorageAPI, mockTxStateAPI, nil),
+			args:      args{},
+			expErr:    errors.New("account address must be valid"),
+		},
+		{
+			name:      "Found",
+			sysModule: NewSystemModule(nil, nil, mockCoreAPI, mockStorageAPI, mockTxStateAPI, nil),
+			args: args{
+				req: &StringRequest{String: "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"},
+			},
+			exp: U64Response(4),
+		},
+		{
+			name:      "Not found",
+			sysModule: NewSystemModule(nil, nil, mockCoreAPI, mockStorageAPI, mockTxStateAPI, nil),
+			args: args{
+				req: &StringRequest{String: "5FrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"},
+			},
+			exp: U64Response(3),
+		},
+		{
+			name:      "GetMetadata Err",
+			sysModule: NewSystemModule(nil, nil, mockCoreAPIErr, mockStorageAPI, mockTxStateAPI, nil),
+			args: args{
+				req: &StringRequest{String: "5FrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"},
+			},
+			expErr: errors.New("getMetadata error"),
+		},
+		{
+			name:      "Magic Number Mismatch",
+			sysModule: NewSystemModule(nil, nil, mockCoreAPIMagicNumMismatch, mockStorageAPI, mockTxStateAPI, nil),
+			args: args{
+				req: &StringRequest{String: "5FrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"},
+			},
+			expErr: errors.New("magic number mismatch: expected 0x6174656d, found 0xe03056ea"),
+		},
+		{
+			name:      "GetStorage Err",
+			sysModule: NewSystemModule(nil, nil, mockCoreAPI, mockStorageAPIErr, mockTxStateAPI, nil),
+			args: args{
+				req: &StringRequest{String: "5FrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"},
+			},
+			expErr: errors.New("getStorage error"),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			res := U64Response(0)
+			sm := tt.sysModule
+			err := sm.AccountNextIndex(tt.args.r, tt.args.req, &res)
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErr.Error())
```

Comment:
Or you could keep `tt.expErr` as string so that you don't have to convert it to string every time you compare. I think this would apply to a lot of other tests.

Reply:
Empty string error doesn't necessarily mean nil error, so that combines a `wantErr bool` + `errMsg string` in one test field.